lib/codegen/meta/src/srcgen.rs

```diff
@@ -12,24 +12,37 @@ use error;
 
 static SHIFTWIDTH: usize = 4;
 
-struct _IndentedScope {
-    fmt: Formatter,
-    after: Option<String>,
+macro_rules! indent {
+    ($fmt:expr, $($middle:tt)*) => {
+        $fmt.indent_push();
+        $($middle)*
+        $fmt.indent_pop();
+    }
 }
```

Comment:
Right, I wasn't even sure of the utility myself. It would have been nice to have a structure IndentedScope, but there's a lifetime issue in there: if we want the `drop()` method to pop by one indent level, then the formatter needs to be mutable in the IndentedScope struct; which means it can't be mutably borrowed between the start of the IndentedScope and the time it's dropped:

```rust
{
  let scope = IndentedScope::new(fmt);
  fmt.write(...); // can't mutate since it's mutably borrowed by scope right above
}
```

So I got back to using this macro instead, which is a small bit nicer than not having it.
If anybody has a better idea for the design of this thing, please chime in!

Reply:
Interesting idea -- I think I like it. As a lambda, it should be easier to step through in a debugger than a macro.